### ISSUE TITLE
Translation validation

### ISSUE THREAD

**Issue by mpazik** [lingui/js-lingui#591]
In our team, we have a problem with the verification of translations. Sometimes translators do a typo in variable names or forgot to close a tag or mess up tags orders and nesting.
If there is a problem in translation js lingui in production mode throws a fatal exception as it can not match variables to the translation and the page is not rendered. Which is really bad.

Right now we have a test parse source and translation for each language and compares but it's not ideal.
It would be nice to have a built-in tool in js lingui CLI, something like `lingui verify` or `lingui compile --check` or be part of `lingui compile --strict`

I could contribute with PR adding this functionality. 
Please let me know if there is some other way to validate the translations that I missed and if not what is preferred way of implementing it from the CLI perspective :)

This ticket could be related to
https://github.com/lingui/js-lingui/issues/397
https://github.com/lingui/js-lingui/issues/4

**Comment by tricoder42** [comment]
Hey @mpazik, I would love to see this feature in Lingui.

1. How do you validate typos in message arguments? We could extend macros by adding "expected" arguments into message metadata and use it for validation. This would be useful e.g. when you use custom IDs, which doesn't include placeholders:

```
<Trans id="Label / Hello">Hello ${name}</Trans>
//<Trans id="Label / Hello" message="Hello {name}" meta={{ args: ['name'] }} />
```

2. Validation of HTML should be possible without other changes I guess.

Please let me know if you're still interested. If so, please use `next` branch as a base for your PR. It'll be released very soon (in matter of days or 1-2 weeks)

**Comment by theKashey** [comment]
(we are on one team) Right now we just checking that all required placeholders exists. Nothing more. That solves 99% of problems.
The check is separated per `{}` and `<>` placeholders, so they cannot change their types.

**Comment by tricoder42** [comment]
> all required placeholders exists

How do you know what placeholders the message uses? You have them in message id?

```js
{
  // if placeholders are included in message, then it's easy
  "Hello {name}": "Salut {nom}",
  
  // if you use custom message Ids, then it's not obvious unless you check the code 
  "Label / Greetings": "Salut {nom}"
}
```

**Comment by mpazik** [comment]
@tricoder42 We load the source file (extracted translations) to memory and then compare translation agains it.
Not sure if loading a file to memory is an issue. Our source file have less then 100KB but it think it should be fine up to like 100MB which is a lot of text. Not sure if there are such cases.

Regarding what exactly we do check:
- all declared ICU variables in translation exists in the source (translation can omit unused variables.
- tags needs to have same structure as in the source file. That means
   - all of them are required
   - nesting of components have to be the same as in the source (`<0>Hello<1/><0>`)
   - the order of components can be different
   - if the tag is closed on the declaration (`<0/>`) in the source, companion tag is not allowed in translation also if the source tag had text, tags in translations also reqruie text.

I could came up within few days with an initial PR so we could disuss exacat checks, the CLI API and implementation.

**Comment by theKashey** [comment]
> How do you know what placeholders the message uses? You have them in message id?

We are parsing the message format and checking that all placeholders used in translation were provided for the original translation. The secondary check, like are they are really provided to the `Trans`(a problem for `react-int`) is not required, as long as there is no way you can "forget" to provide it using standard semantic. However, `intl._` is not secured. 

> tags needs to have same structure as in the source file. That means...

Actually, that broke Japanise translation and right now is disabled. I was thinking about this for a while, and:
- all of them are required
> you don't have to use all placeholders. Some language might be more "compact", some less. Ideally you should write the original message not in English, to be able to provide all details needed for different languages (like gender, or other stuff [FTB](https://github.com/facebook/fbt), for example, provides)

- nesting of components have to be the same as in the source
> I was trying to find a way how this could break anything, and look like it cannot.

- if the tag is closed on the declaration
> that depends on what those tags are, and why there were left there.

---

Ideally, translation validation should prevent runtime errors. Nothing more. The result might looks a bit odd, even broken, but what if it is intentional?
It shall be no more that linting.

**Comment by mpazik** [comment]
I see, the checks were relaxed recently in our repo, I haven't fetch master of that project since few weeks. 
I agree that they sould only catch issues that would break page rendering.

**Comment by mpazik** [comment]
I did some more investigation and I changed my mind. 
I think that any typo in the translation shouldn't break the page in the runtime, perhaps it could cause a giltch in the dispalyed text like variable would be ignored or an extra tag would be displayed like `Hello <1>`.

Some of the issues that used to break a page don't break it any more. 
It will rise a bug for the remaining like using tag that does not exists or type in ICU types.

In the PR I'm going to introduce `lingui verify` that will check for obvious translator errors that would cause giltches in the UI (including current critial erros).
Things that could be made on purpouse like ignoring variable, different component order or structure, will be allowed.
